hibernate-core/src/main/java/org/hibernate/type/spi/TypeConfiguration.java

```diff
@@ -640,6 +640,9 @@ public ArrayCacheKey(SqmExpressible<?>[] components) {
 
 		@Override
 		public boolean equals(Object o) {
+			if (!(o instanceof ArrayCacheKey)) {
```

Comment:
`ArrayCacheKey` is a private class and we ensured this method is never called with an argument of a different type, so this check is unnecessary.